File: packages/parallax_cli/src/commands/edit.ts
// Edit command and all its subcommands (image, video).
// Action handlers follow: validate → resolveModelsDir → buildArgs → spawnPipeline.

import { existsSync } from "fs";
import { Command } from "commander";
import { readConfig } from "../config";
import { spawnPipeline } from "../runner";
import { resolveModelsDir } from "../utils";
import { getModels, getScript } from "../models/registry";
import { buildEditImageArgs, type EditImageOpts } from "../models/image";
import { modelsFooter, validateModel, notImplemented } from "../utils";

export function registerEdit(program: Command): void {
  const edit = program
    .command("edit")
    .description("Edit existing media using a text prompt")
    .usage("<media> [options]");

  edit
    .command("image")
    .description("Edit an image using a text prompt")
    .requiredOption("--model <name>", `Model to use (choices: ${getModels("edit", "image").join(", ")})`)
    .requiredOption("--prompt <text>", "Text prompt describing the desired edits")
    .requiredOption("--input <path>", "Path to the input image file")
    .option("--subject-image <path>", "Subject reference image (flux_9b_kv only)")
    .option("--width <pixels>", "Image width in pixels", "1024")
    .option("--height <pixels>", "Image height in pixels", "1024")
    .option("--steps <n>", "Number of sampling steps (qwen: auto 4 with LoRA, 40 without)")
    .option("--cfg <value>", "CFG guidance scale (qwen: auto 1.0 with LoRA, 4.0 without)")
    .option("--seed <n>", "Random seed for reproducibility")
    .option("--output <path>", "Output file path", "output.png")
    .option("--image2 <path>", "Second input image (qwen only)")
    .option("--image3 <path>", "Third input image (qwen only)")
    .option("--no-lora", "Disable LoRA (qwen only)")
    .option("--models-dir <path>", "Models directory (overrides PYCOMFY_MODELS_DIR)")
    .addHelpText("after", modelsFooter("edit", "image"))
    .action(async (opts) => {
      validateModel("edit", "image", opts.model);
      const script = getScript("edit", "image", opts.model);
      if (!script) notImplemented("edit", "image", opts.model);
      if (!existsSync(opts.input)) {
        console.error(`Error: input file not found: ${opts.input}`);
        process.exit(1);
      }
      if (opts.model === "flux_9b_kv") {
        if (!opts.subjectImage) {
          console.error("Error: --subject-image is required for flux_9b_kv");
          process.exit(1);
        }
        if (!existsSync(opts.subjectImage)) {
          console.error(`Error: subject-image file not found: ${opts.subjectImage}`);
          process.exit(1);
        }
      }
      const modelsDir = resolveModelsDir(opts.modelsDir);
      const editOpts: EditImageOpts = {
        model:        opts.model,
        prompt:       opts.prompt,
        input:        opts.input,
        width:        opts.width,
        height:       opts.height,
        steps:        opts.steps,
        cfg:          opts.cfg,
        seed:         opts.seed,
        output:       opts.output,
        subjectImage: opts.subjectImage,
        image2:       opts.image2,
        image3:       opts.image3,
        noLora:       !!opts.noLora,
      };
      const args = buildEditImageArgs(editOpts, modelsDir);
      await spawnPipeline(script, args, readConfig());
    });

  edit
    .command("video")
    .description("Edit a video using a text prompt")
    .requiredOption("--model <name>", `Model to use (choices: ${getModels("edit", "video").join(", ")})`)
    .requiredOption("--prompt <text>", "Text prompt describing the desired edits")
    .requiredOption("--input <path>", "Path to the input video file")
    .option("--width <pixels>", "Output video width in pixels")
    .option("--height <pixels>", "Output video height in pixels")
    .option("--length <frames>", "Number of frames to generate")
    .option("--steps <n>", "Number of sampling steps", "30")
    .option("--cfg <value>", "CFG guidance scale", "6")
    .option("--seed <n>", "Random seed for reproducibility")
    .option("--output <path>", "Output file path", "output.mp4")
    .option("--models-dir <path>", "Models directory (overrides PYCOMFY_MODELS_DIR)")
    .addHelpText("after", modelsFooter("edit", "video"))
    .action((opts) => {
      validateModel("edit", "video", opts.model);
      notImplemented("edit", "video", opts.model);
    });
}

File: cli/commands/edit.py
"""``parallax edit`` subcommand group — image editing."""

from __future__ import annotations

from typing import Annotated, Optional

import typer

from cli._io import resolve_models_dir, save_image
from cli._runners import edit_image

app = typer.Typer(help="Edit existing media using a model and prompt.")


@app.command("image")
def edit_image_cmd(
    model: Annotated[str, typer.Option("--model", help=f"Model. Choices: {', '.join(edit_image.MODELS)}.")],
    prompt: Annotated[str, typer.Option("--prompt", help="Editing instruction or text prompt.")],
    input: Annotated[str, typer.Option("--input", help="Path to the input image file.")],
    width: Annotated[Optional[int], typer.Option("--width", help="Output width in pixels.")] = None,
    height: Annotated[Optional[int], typer.Option("--height", help="Output height in pixels.")] = None,
    steps: Annotated[Optional[int], typer.Option("--steps", help="Sampling steps.")] = None,
    cfg: Annotated[Optional[float], typer.Option("--cfg", help="CFG guidance scale.")] = None,
    seed: Annotated[int, typer.Option("--seed", help="Random seed.")] = 0,
    output: Annotated[str, typer.Option("--output", help="Output file path.")] = "output.png",
    models_dir: Annotated[Optional[str], typer.Option("--models-dir", help="Models directory.")] = None,
    subject_image: Annotated[Optional[str], typer.Option("--subject-image", help="Subject reference image (flux_9b_kv only).")] = None,
    image2: Annotated[Optional[str], typer.Option("--image2", help="Optional second reference image (qwen only).")] = None,
    image3: Annotated[Optional[str], typer.Option("--image3", help="Optional third reference image (qwen only).")] = None,
    no_lora: Annotated[bool, typer.Option("--no-lora", help="Disable Lightning LoRA (qwen only).")] = False,
    async_mode: Annotated[bool, typer.Option("--async", help="Queue job, return ID immediately.")] = False,
) -> None:
    """Edit an image using an instruction prompt and a reference image."""
    if model not in edit_image.MODELS:
        typer.echo(f"Error: unknown model '{model}'. Choices: {', '.join(edit_image.MODELS)}.", err=True)
        raise typer.Exit(code=1)
    if async_mode:
        from cli._async import run_async
        run_async(action="edit", media="image", model=model,
                  args={"prompt": prompt, "input": input, "width": width, "height": height,
                        "steps": steps, "cfg": cfg, "seed": seed, "output": output,
                        "models_dir": models_dir, "subject_image": subject_image,
                        "image2": image2, "image3": image3, "no_lora": no_lora})
        return
    from pathlib import Path as _Path
    if not _Path(input).is_file():
        typer.echo(f"Error: input file not found: {input}", err=True)
        raise typer.Exit(code=1)
    if model == "flux_9b_kv":
        if subject_image is None:
            typer.echo("Error: --subject-image is required for flux_9b_kv.", err=True)
            raise typer.Exit(code=1)
        if not _Path(subject_image).is_file():
            typer.echo(f"Error: subject image not found: {subject_image}", err=True)
            raise typer.Exit(code=1)
    mdir = resolve_models_dir(models_dir)
    w = width  or int(edit_image.default(model, "width",  1024))
    h = height or int(edit_image.default(model, "height", 1024))
    s = steps  or int(edit_image.default(model, "steps",  30))
    c = cfg    if cfg is not None else float(edit_image.default(model, "cfg", 3.5))
    try:
        from PIL import Image as _PIL
        img      = _PIL.open(input).convert("RGB")
        subj     = _PIL.open(subject_image).convert("RGB") if subject_image else None
        img2_pil = _PIL.open(image2).convert("RGB") if image2 else None
        img3_pil = _PIL.open(image3).convert("RGB") if image3 else None
        images = edit_image.RUNNERS[model](
            mdir=mdir, img=img, prompt=prompt, w=w, h=h, s=s, c=c, seed=seed,
            subject_img=subj, img2=img2_pil, img3=img3_pil, no_lora=no_lora,
        )
        typer.echo(save_image(images, output))
    except (typer.Exit, SystemExit):
        raise
    except Exception as exc:
        typer.echo(f"Error: {exc}", err=True)
        raise typer.Exit(code=1) from exc
